File: solutions/lua/dnd-character/1/dnd-character.lua
local function modifier(input) return (input - 10) // 2 end
local function ability()
  local dice = { math.random(1, 6), math.random(1, 6), math.random(1, 6), math.random(1, 6) }
  table.sort(dice, function(a, b) return a < b end)
  return dice[1] + dice[2] + dice[3]
end

local Character = {}
function Character:new(name)
  local constitution = ability()
  return {
    name = name,
    strength = ability(),
    dexterity = ability(),
    constitution = constitution,
    intelligence = ability(),
    wisdom = ability(),
    charisma = ability(),
    hitpoints = 10 + modifier(constitution),
  }
end

return { Character = Character, ability = ability, modifier = modifier }

File: solutions/ruby/dnd-character/1/dnd_character.rb
class DndCharacter
  attr_reader :constitution, :strength, :dexterity, :intelligence, :wisdom, :charisma, :hitpoints

  def self.modifier(input) = (input - 10) / 2
  def initialize
    rolls = 6.times.map { 4.times.map { Random.new.rand(1...6) }.sort[-3..].sum }
    @constitution, @strength, @dexterity, @intelligence, @wisdom, @charisma = *rolls
    @hitpoints = 10 + self.class.modifier(@constitution)
  end
end
